lib/streamlit/connections/snowpark_connection.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: We won't always be able to import from snowflake.snowpark.session so need the
+# `type:ignore` comment below, but that comment will explode if `warn-unused-ignores` is
+# turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+import configparser
+import os
+import threading
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Iterator, Optional, Union, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account", "user"}
+_DEFAULT_CONNECTION_FILE = "~/.snowsql/config"
+
+
+def _load_from_snowsql_config_file() -> Dict[str, Any]:
+    """Loads the dictionary from snowsql config file."""
+    snowsql_config_file = os.path.expanduser(_DEFAULT_CONNECTION_FILE)
+    if not os.path.exists(snowsql_config_file):
+        return {}
+
+    config = configparser.ConfigParser(inline_comment_prefixes="#")
+    config.read(snowsql_config_file)
+
+    conn_params = config["connections"]
+    conn_params = {k.replace("name", ""): v.strip('"') for k, v in conn_params.items()}
+    if "db" in conn_params:
+        conn_params["database"] = conn_params["db"]
+        del conn_params["db"]
+    return conn_params
+
+
+class Snowpark(BaseConnection["Session"]):
+    _default_connection_name = "snowpark"
+
+    def __init__(self, connection_name: str = "default", **kwargs) -> None:
+        self._lock = threading.RLock()
+
+        # Grab the lock before calling BaseConnection.__init__() so that we can guarantee
+        # thread safety when the parent class' constructor initializes our connection.
+        with self._lock:
+            super().__init__(connection_name, **kwargs)
+
+    # TODO(vdonato): Teach the .connect() method how to automagically connect in a SiS
+    # runtime environment.
+    def connect(self, **kwargs) -> "Session":
```

Comment:
I wonder if we could be more explicit with the config options here? E.g.:

`def connect(self, account: Optional[str], user: Optional[str], ..., **kwargs)` + a good docstring explaining the options.

It is probably enough to only cover the most common options, and leave the rest to `kwargs`. By not doing this, we might require the user to figure out the config options on their own by finding them in the underlying library. Which is a bit contrary to the promise of simplifying connections. Of course, just putting info about the config options into our documentation could also be an option.

Having this somehow more explicit could also be very useful for autogenerating the connection wizard later. 

Linking @sfc-gh-jcarroll here since this is also bridging a bit into product.

Reply:
I agree covering the common options is a good idea. Not necessarily a ship blocker but definitely something to do before GA and it'll be nicer if it's done now.

Similar to the practice I described [here](https://www.notion.so/snowflake-corp/BaseConnection-class-API-4c495df9e8cb4958a2da942e00a8d3c6?pvs=4#9fb75ada03b44b0191846d35b77379f9).

Reply:
ack -- I have a separate ticket filed to figure out how to merge what's found in files / secrets / kwargs to the object passed to connection params, so I'll address this as part of that work item